### PR TITLE
Stp intercept tactic

### LINKED ISSUE #856: Intercept
Uses estimated ball end position from play. 
Calculate the closest interception point on the line the ball will move on.
Move there
Turn the robot to receive the ball.
Turn on dribbler

### PR DESCRIPTION
**Pull request by timovdk**
Implemented the Intercept tactic. There should be some extra calculations done when deciding the intercept point, since now the ball position is used. This is up to control though.
Closes #856

### CODE REVIEW

File: include/roboteam_ai/stp/new_tactics/Intercept.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#ifndef RTT_INTERCEPT_H
+#define RTT_INTERCEPT_H
+
+#include "stp/Tactic.h"
+
+namespace rtt::ai::stp::tactic {
+class Intercept : public Tactic {
+   public:
+    Intercept();
+    virtual bool isEndTactic() noexcept;
```
Comment (jordiluong) [inline include/roboteam_ai/stp/new_tactics/Intercept.h]:
```suggestion
```

File: include/roboteam_ai/stp/new_tactics/Intercept.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    bool isTacticFailing(const StpInfo &info) noexcept override;
+
+    /**
+     * Reset tactic when robot position is not close enough to the target position for receiving
+     * @param info Info
+     * @return True if robot position is not close enough to the target position
+     */
+    bool shouldTacticReset(const StpInfo &info) noexcept override;
+
```
Comment (jordiluong) [inline include/roboteam_ai/stp/new_tactics/Intercept.h:51]:
```suggestion

bool isEndTactic() noexcept override;

```

File: src/stp/new_tactics/Intercept.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+bool Intercept::isTacticFailing(const StpInfo& info) noexcept {
+    // If the ball doesn't move, the robot can't intercept
+    return info.getBall()->get()->getVelocity().length() <= Constants::BALL_STILL_VEL();
+}
+
+bool Intercept::shouldTacticReset(const StpInfo& info) noexcept {
+    // If the robot is not on the velocity vector of the ball, reset this tactic to move towards this line again
+    return info.getRobot()->get()->getPos() - info.getBall()->get()->getVelocity() != info.getBall()->get()->getVelocity() - info.getRobot()->get()->getPos();
```
Comment (jordiluong) [inline src/stp/new_tactics/Intercept.cpp]: Maybe add small error margin

**Review by jordiluong — APPROVED** [review]
Looks good

File: src/stp/new_tactics/Intercept.cpp (hunk 3 above)
Comment (timovdk) [reply src/stp/new_tactics/Intercept.cpp]: Oh I was debating what margin to use and in the end just forgot about it woops

File: include/roboteam_ai/stp/Play.hpp [hunk 4]
```diff
@@ -82,7 +82,7 @@ class Play {
     /**
      * The roles, constructed in ctor of a play
      */
-    std::array<std::unique_ptr<Role>, rtt::ai::Constants::ROBOT_COUNT()> roles;
+    std::array<std::unique_ptr<Role>, /*rtt::ai::Constants::ROBOT_COUNT()*/2> roles;
```
Comment (RobotJente) [inline include/roboteam_ai/stp/Play.hpp]:
```suggestion
    std::array<std::unique_ptr<Role>, rtt::ai::Constants::ROBOT_COUNT()> roles;
```

File: src/stp/new_plays/TestPlay.cpp [hunk 5]
```diff
@@ -27,16 +27,16 @@ void TestPlay::assignRoles() noexcept {
     Dealer::DealerFlag closeToTheirGoalFlag(DealerFlagTitle::CLOSE_TO_THEIR_GOAL, DealerFlagPriority::MEDIUM_PRIORITY);
 
     flagMap.insert({"test_role_0", {closeToBallFlag}});
-    flagMap.insert({"test_role_1", {closeToTheirGoalFlag}});
-    flagMap.insert({"test_role_2", {closeToTheirGoalFlag, closeToBallFlag}});
+    flagMap.insert({"kicker", {closeToTheirGoalFlag}});
+/*    flagMap.insert({"test_role_2", {closeToTheirGoalFlag, closeToBallFlag}});
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
    flagMap.insert({"test_role_2", {closeToTheirGoalFlag, closeToBallFlag}});
```

File: src/stp/new_plays/TestPlay.cpp [hunk 6]
```diff
@@ -27,16 +27,16 @@ void TestPlay::assignRoles() noexcept {
...
     flagMap.insert({"test_role_3", {closeToBallFlag}});
     flagMap.insert({"test_role_4", {closeToTheirGoalFlag}});
     flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}});
     flagMap.insert({"test_role_6", {closeToBallFlag}});
     flagMap.insert({"test_role_7", {closeToTheirGoalFlag}});
     flagMap.insert({"test_role_8", {closeToTheirGoalFlag, closeToBallFlag}});
     flagMap.insert({"test_role_9", {closeToBallFlag}});
-    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});
+    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});*/
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});
```

File: src/stp/new_plays/TestPlay.cpp [hunk 7]
```diff
@@ -8,11 +8,11 @@
 namespace rtt::ai::stp {
 
 TestPlay::TestPlay() {
-    roles = std::array<std::unique_ptr<Role>, rtt::ai::Constants::ROBOT_COUNT()>{
-        std::make_unique<TestRole>(TestRole("test_role_0")), std::make_unique<TestRole>(TestRole("test_role_1")), std::make_unique<TestRole>(TestRole("test_role_2")),
+    roles = std::array<std::unique_ptr<Role>, 2/*rtt::ai::Constants::ROBOT_COUNT()*/>{
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
    roles = std::array<std::unique_ptr<Role>, rtt::ai::Constants::ROBOT_COUNT()>{
```

File: src/stp/new_plays/TestPlay.cpp [hunk 8]
```diff
@@ -8,11 +8,11 @@
...
+        std::make_unique<TestRole>(TestRole("test_role_0")), std::make_unique<TestRole>(TestRole("kicker"))/*, std::make_unique<TestRole>(TestRole("test_role_2")),
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
        std::make_unique<TestRole>(TestRole("test_role_0")), std::make_unique<TestRole>(TestRole("kicker")), std::make_unique<TestRole>(TestRole("test_role_2")),
```

File: src/stp/new_plays/TestPlay.cpp [hunk 9]
```diff
@@ -8,11 +8,11 @@
...
         std::make_unique<TestRole>(TestRole("test_role_3")), std::make_unique<TestRole>(TestRole("test_role_4")), std::make_unique<TestRole>(TestRole("test_role_5")),
         std::make_unique<TestRole>(TestRole("test_role_6")), std::make_unique<TestRole>(TestRole("test_role_7")), std::make_unique<TestRole>(TestRole("test_role_8")),
-        std::make_unique<TestRole>(TestRole("test_role_9")), std::make_unique<TestRole>(TestRole("test_role_10"))};
+        std::make_unique<TestRole>(TestRole("test_role_9")), std::make_unique<TestRole>(TestRole("test_role_10"))*/};
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
        std::make_unique<TestRole>(TestRole("test_role_9")), std::make_unique<TestRole>(TestRole("test_role_10"))};
```

File: src/stp/new_tactics/Intercept.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    return true;
+}
+
+void Intercept::onInitialize() noexcept {}
+
+void Intercept::onUpdate(const Status& status) noexcept {
+    // Keep executing Rotate skill
+    if (skills.current_num() == skills.total_count()) {
+        skills.skip_n(-1);
+    }
```
Comment (RobotJente) [inline src/stp/new_tactics/Intercept.cpp:29]: What is this for?

File: src/stp/new_tactics/Intercept.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    return skillStpInfo;
+}
+
+bool Intercept::isTacticFailing(const StpInfo& info) noexcept {
+    // If the ball doesn't move, the robot can't intercept
+    return info.getBall()->get()->getVelocity().length() <= Constants::BALL_STILL_VEL();
+}
+
+bool Intercept::shouldTacticReset(const StpInfo& info) noexcept {
+    // If the robot is not on the velocity vector of the ball, reset this tactic to move towards this line again
```
Comment (RobotJente) [inline src/stp/new_tactics/Intercept.cpp]: Fix

**Review by RobotJente — APPROVED** [review]

**Pull request merged**